### ISSUE TITLE
Datastore sync is blocking main thread

### ISSUE THREAD

**Issue by alionthego** [aws-amplify/amplify-swift#1409]
I have begun noticing when my app wakes from background after being in background several minutes, my UI scrolling is momentarily choppy until the database has finished syncing and Database is Ready.

How can I ensure that the syncing is done off main thread?  At the moment I am using the following code:

```
class ViewModel: ObservableObject {
    
    @Published var posts: [Post] = []
    
    var databaseReady = false
    var syncQueriesReady = true
    
    init() {
        
        self.query()
        
        let _ = Amplify.Hub.listen(to: .dataStore) { event in
            DispatchQueue.main.async {
                if event.eventName == HubPayload.EventName.DataStore.syncQueriesStarted {
                    print("syncQueriesStarted")
                    self.syncQueriesReady = false
                }
                if event.eventName == HubPayload.EventName.DataStore.syncQueriesReady {
                    print("syncQueriesReady")
                    self.syncQueriesReady = true
                }
                if event.eventName == HubPayload.EventName.DataStore.ready {
                    print("syncDatabaseReady")
                    self.databaseReady = true
                    self.query()
                }
                if event.eventName == HubPayload.EventName.DataStore.outboxMutationEnqueued {
                    self.query()
                }
                if event.eventName == HubPayload.EventName.DataStore.syncReceived && self.databaseReady && self.syncQueriesReady {
                    self.query()
                }
            }
        }
    }
    
    private func query() {
        DispatchQueue.global(qos: DispatchQoS.QoSClass.userInitiated).async {
            Amplify.DataStore.query(Post.self, sort: .ascending(Post.keys.name)) {
                switch $0 {
                case .success(let result):
                    DispatchQueue.main.async {
                        self.posts = result
                    }
                case .failure(let error):
                    print("Error listing posts - \(error.localizedDescription)")
                }
            }
            
        }
    }
}

```

Because my actual schema has a lot of relation's, using a publisher is not going to work and neither is just querying the mutated item.  It get's very complicated with relations so I'm just querying the database each time knowing there are a few hundred posts.

But my UI seems to be freezing before the query is called.  It is freezing mostly when app comes out of suspended state while the database is between syncQueriesStarted and syncQueriesReady

**Comment by rawandahmad698** [comment]
Use the background thread, not main!

` DispatchQueue.global(qos: .background).asyncAfter(deadline: .now() + 0.1, execute: {
                    // Your code
                })`

**Comment by ameter** [comment]
That's correct.  And to clarify, you are correct to use DispatchQueue.main.async in the completion handler of the query function to update self.posts.  However, the issue is that in the init function, you should not be using the main thread (e.g. use DispatchQueue.global as rawandahmad698 said) to kick off the query.

**Comment by alionthego** [comment]
Hi.  Very sorry for omitting the threading in the original post.  I have been dispatching the query to the background as the updated code shows in the query function.  And this problem exists.  Is there anything in the code above that should be causing this significantly degraded performance when coming out of background after being in background for a significant period of time?

**Comment by lawmicha** [comment]
Hi @alionthego, could you show us the latest updated code to see if we can identify why the problem still exists?

**Comment by alionthego** [comment]
Hi.  I have just discovered that the blocking of the main thread was caused by another viewModel and not this one.  

Although it is a similar issue.  I was listening for HubPayload.EventName.DataStore.syncReceived and updating model.  The issue in that class was that I was not disabling it during database sync causing thousands of calls.  

I would appreciate so so much if you guys can have a working example that lets people know how we are supposed to perform simple tasks such as updating models when data changes.  When to disable and re enable subscriptions or hub listeners.  

When does the datastore perform a sync after being in background.  Why does the database not always sync after being in the background and coming back to foreground?  Is there a timeout?  Will data be out of sync if coming out of background and no sync is performed?  Do I need to enable background mode in my app?

I have dozens of more issues which are left for us to guess and try different approaches to see what works.  It's a huge waste of time.  If this is really a proper solution for building apps then kindly build one and let us see how to do it properly.  Use your blogs / comments example.  Thanks
